File: tests/bench/dict_iter/dict_iter.cpp
// Peer port of dict_iter.jai. Must print byte-identical output.
#include <cstdio>
#include <cstdlib>
#include <string>
#include <unordered_map>

int main() {
    const char *env = std::getenv("BENCH_LEVEL");
    std::string level = env ? env : "hard";
    int scale = level == "easy" ? 20 : level == "medium" ? 5 : 1;
    const int KEYS = 20000;
    int reps = 120 / scale;
    if (reps < 1) reps = 1;

    std::unordered_map<std::string, long long> d;
    d.reserve(KEYS * 2);
    for (int i = 0; i < KEYS; i++) d["k" + std::to_string(i)] = i;

    long long total = 0;
    long long seen = 0;
    for (int r = 0; r < reps; r++) {
        total = 0;
        for (const auto &kv : d) total += kv.second;
        seen = 0;
        for (const auto &kv : d) {
            if (kv.first.size() > 4 && kv.second % 2 == 0) seen++;
        }
    }
    std::printf("%lld\n%lld\n", total, seen);
    return 0;
}

File: tests/bench/dict_iter/dict_iter.py
import os

LEVEL = os.environ.get("BENCH_LEVEL", "hard")
SCALE = 20 if LEVEL == "easy" else 5 if LEVEL == "medium" else 1
KEYS = 20000
REPS = max(120 // SCALE, 1)


def build(n):
    return {f"k{i}": i for i in range(n)}


def sum_values(d):
    total = 0
    for _k, v in d.items():
        total += v
    return total


def count_long_keys(d):
    seen = 0
    for k, v in d.items():
        if len(k) > 4 and v % 2 == 0:
            seen += 1
    return seen


def main():
    d = build(KEYS)
    total = 0
    seen = 0
    for _ in range(REPS):
        total = sum_values(d)
        seen = count_long_keys(d)
    print(total)
    print(seen)


main()
